client/fetch_tagged_results_accumulator.go

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/m3db/m3cluster/shard"
+	"github.com/m3db/m3db/encoding"
+	"github.com/m3db/m3db/generated/thrift/rpc"
+	"github.com/m3db/m3db/storage/index"
+	"github.com/m3db/m3db/topology"
+	xerrors "github.com/m3db/m3x/errors"
+)
+
+type fetchResultsAccumulator interface {
+	// Reset resets the accumulator for use.
+	Reset(
+		startTime time.Time,
+		endTime time.Time,
+		topoMap topology.Map,
+		majority int,
+		consistencyLevel topology.ReadConsistencyLevel)
+
+	// Accumulate add the provided result to the accumulator.
+	Accumulate(result interface{}, err error) (done bool, accumErr error)
+
+	// AsEncodingSeriesIterators converts the underlying response into an encoding.SeriesIterators
+	AsEncodingSeriesIterators(limit int, pools fetchTaggedPools) (
+		iters encoding.SeriesIterators, exhaustive bool, err error)
+
+	// AsIndexQueryResults converts the underlying response into an index.QueryResults
+	AsIndexQueryResults(limit int, pools fetchTaggedPools) (index.QueryResults, error)
+
+	// Clear releases any held resources.
+	Clear()
```

Comment:
Is this conceptually different than Close?

Reply:
It's similar to the degree that both Close() and Clear() release internal state.

The difference is close() releases the object that it's called upon back into a pool. Clear() does not.  This is because the user of this (fetchState) retains the reference to the accumulator, and it's pooled at the fetchState level, not the the accumulator level.